huixiangdou/Readme.md

```diff
@@ -0,0 +1,381 @@
+
+![](imgs/env.png)
+
+后续教程所有操作都需要在该环境下进行。
+
+### 1.2 下载基础文件
+
+复制茴香豆所需模型，为了减少下载和huggingface登录问题，所有作业和教程涉及的模型都已经存放在共享文件中，大家根据自己机器配置选择合适的模型复制。
+
+```bash
+# 创建模型文件夹
+cd /root && mkdir models
+
+# 复制BCE模型
+ln -s /root/share/new_models/maidalun1020/bce-embedding-base_v1 /root/models/bce-embedding-base_v1
+ln -s /root/share/new_models/maidalun1020/bce-reranker-base_v1 /root/models/bce-reranker-base_v1
+
+# 复制大模型参数（下面的模型，根据作业进度和任务进行选择一个就行）
+## 基础作业 (8G显存)
+ln -s /root/share/new_models/Shanghai_AI_Laboratory/internlm2-chat-1_8b /root/models/internlm2-chat-1_8b
+
+## 进阶作业(20G显存)
+ln -s /root/share/new_models/Shanghai_AI_Laboratory/internlm2-chat-7b /root/models/internlm2-chat-7b
+
+```
+
+### 1.3 下载安装茴香豆（豆哥）
+
+安装系统和茴香豆解析 **Word** 文件依赖。
+
+```
+## 安装系统必要组件
+apt update
+apt -y install python-dev python
+
+
+# 安装解析 .doc .docx 必须依赖
+apt -y install  libxml2-dev libxslt1-dev antiword unrtf poppler-utils pstotext tesseract-ocr flac ffmpeg lame libmad0 libsox-fmt-mp3 sox libjpeg-dev swig libpulse-dev 
+```
+> 注意！由于 **InternLM Studio** 限制，每次重启后上述命令都需要重新执行，重新安装。
+
+从茴香豆官方仓库下载茴香豆。
+
+```bash
+cd /root
+# 下载 repo
+git clone https://github.com/internlm/huixiangdou && cd huixiangdou
+git checkout 447c6f7e68a1657fce1c4f7c740ea1700bde0440
+
+# python requirements
+pip install -r requirements.txt
+pip install protobuf # internlm2
+```
+
+
+
+### 1.4 端口映射（可选）
+
+后续如果想通过gradio搭建web应用demo，请跟随下面的步骤将服务器端口映射到本地端口：
+首先我们需要配置一下本地的 SSH Key ，我们这里以Windows为例。
+
+1. 查询服务器端口和密码（图中端口示例为 38374）：
+
+![](imgs/check_port.png)
+
+2. 在本地打开命令行.
+    - Windows 使用快捷键组合 `Windows + R`（Windows 即开始菜单键）打开指令界面，并输入命令 `Powershell`，按下回车键
+    - Mac 用户直接找到并打开`终端`
+    - Ubuntu 用户使用快捷键组合 `ctrl + alt + t`
+
+3. 在命令行中输入如下命令，命令行会提示输入密码：
+```
+ssh -CNg -L 7860:127.0.0.1:7860 root@ssh.intern-ai.org.cn -p 38074
+```
+4. 复制服务器密码到命令行中，按回车，建立服务器到本地到端口映射。
+
+![Alt text](imgs/port_psw.png)
+
+
+## 2 使用茴香豆
+
+### 2.1 修改配置文件
+
+用已下载模型的路径替换 `/root/huixiangdou/config.ini` 文件中的模型，如下图所示：
+
+![](imgs/model_path.png)
+
+> 基础作业 
+
+`local_llm_path = "/root/models/internlm2-chat-1_8b"`
+
+基础作业可以参考这个文件修改 [config.ini](/config.ini)。
+
+> 进阶作业
+
+`local_llm_path = "/root/models/internlm2-chat-7b"`
+
+### 2.2 收集本地数据库
+
+创建本地知识数据库。本示例中，使用OpenMMLab 的 MMPose 文档作为数据库，打造一个知识问答助手。
+
+```bash
+cd /root/huixiangdou && mkdir repodir
+git clone https://github.com/open-mmlab/mmpose --depth=1 repodir/mmpose
+```
+
+修改文件 `/root/huixiangDou/resource/good_questions.json` 和 `/root/huixiangDou/resource/bad_questions.json` ，添加或删除希望模型回答和模型拒绝回答的问题列表。
+
+> 注意！默认列表根据 MMPose 创建，如果使用自己的数据库，对应接受和拒绝回答问题要根据知识库领域进行修改。
+
+```
+# 默认知识库接受问题类型
+"mmpose中怎么调用mmyolo接口",
+"mmpose实现姿态估计后怎么实现行为识别",
+...
+
+# 默认知识库拒绝问题类型（与主题无关或闲聊）
+"nihui 是谁",
+"具体在哪些位置进行修改？",
+"你是谁？",
+"1+1",
+...
+```
+
+提取知识库特征，创建向量数据库。数据库向量化的过程应用到了langchain的相关模块，默认嵌入和重排序模型调用的网易BCE模型，如果没有在`config.ini`文件中指定本地模型路径，茴香豆将自动从hugging-face下载默认模型。
+
+```bash
+# 创建工作目录
+cd /root/huixiangdou && mkdir workdir 
+
+# 把 repodir 中的数据向量化后保存到 workdir 向量数据库中，便于后续搜索查询使用
+python3 -m huixiangdou.service.feature_store 
```

Comment:
```shell
(InternLM2_Huixiangdou) root@intern-studio-155461:~/huixiangdou# python3 -m huixiangdou.service.feature_store 

/root/.conda/envs/InternLM2_Huixiangdou/lib/python3.10/site-packages/langchain/embeddings/__init__.py:29: LangChainDeprecationWarning: Importing embeddings from langchain is deprecated. Importing from langchain will no longer be supported as of langchain==0.2.0. Please import from langchain-community instead:

`from langchain_community.embeddings import HuggingFaceEmbeddings`.

To install langchain-community run `pip install -U langchain-community`.
  warnings.warn(
/root/.conda/envs/InternLM2_Huixiangdou/lib/python3.10/site-packages/langchain/embeddings/__init__.py:29: LangChainDeprecationWarning: Importing embeddings from langchain is deprecated. Importing from langchain will no longer be supported as of langchain==0.2.0. Please import from langchain-community instead:

`from langchain_community.embeddings import HuggingFaceEmbeddings`.

To install langchain-community run `pip install -U langchain-community`.
  warnings.warn(
/root/.conda/envs/InternLM2_Huixiangdou/lib/python3.10/runpy.py:126: RuntimeWarning: 'huixiangdou.service.feature_store' found in sys.modules after import of package 'huixiangdou.service', but prior to execution of 'huixiangdou.service.feature_store'; this may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
/root/.conda/envs/InternLM2_Huixiangdou/lib/python3.10/site-packages/langchain/embeddings/__init__.py:29: LangChainDeprecationWarning: Importing embeddings from langchain is deprecated. Importing from langchain will no longer be supported as of langchain==0.2.0. Please import from langchain-community instead:

`from langchain_community.embeddings import HuggingFaceEmbeddings`.

To install langchain-community run `pip install -U langchain-community`.
  warnings.warn(
Traceback (most recent call last):
  File "/root/.conda/envs/InternLM2_Huixiangdou/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/root/.conda/envs/InternLM2_Huixiangdou/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/root/huixiangdou/huixiangdou/service/feature_store.py", line 531, in <module>
    cache = CacheRetriever(config_path=args.config_path)
  File "/root/huixiangdou/huixiangdou/service/retriever.py", line 197, in __init__
    config = pytoml.load(f)['feature_store']
  File "/root/.conda/envs/InternLM2_Huixiangdou/lib/python3.10/site-packages/pytoml/parser.py", line 11, in load
    return loads(fin.read(), translate=translate, object_pairs_hook=object_pairs_hook, filename=getattr(fin, 'name', repr(fin)))
  File "/root/.conda/envs/InternLM2_Huixiangdou/lib/python3.10/site-packages/pytoml/parser.py", line 70, in loads
    error('duplicate_tables')
  File "/root/.conda/envs/InternLM2_Huixiangdou/lib/python3.10/site-packages/pytoml/parser.py", line 27, in error
    raise TomlError(msg, pos[0], pos[1], filename)
pytoml.core.TomlError: config.ini(103, 1): duplicate_tables
```

Reply:
fixed
重复粘贴了